### ISSUE TITLE
Message editing and moderation architectures

### ISSUE THREAD

**Issue by cruickshankpg** [ably-labs/fully-featured-scalable-chat-app#59]
This is a proposal for two different message processing architectures that I think it would be interesting/useful to see in FFS.

It's common for chat apps to need to edit, augment or moderate messages. A chat is usually formed around a single shared channel to which all users subscribe for messages. Depending on the use case there are a few different patterns that may be appropriate for extending functionality to allow for better experiences:

##Delayed augmentation

For use cases where some further processing needs to happen on a message, for example a bot looking up address details from a message and providing a maps link. 

A user publishes to the shared chat channel so that the message is immediately shared with all users. The channel also has a rule configured that invokes a cloud function and runs whatever extended functionality is desired and publishes back on the channel. User's apps then augment the original message with this additional message.

The processing for the cloud function is relatively straightforward to be able to process multiple different chat channels as it just publishes messages back onto the channel that they came from.

Benefits:
- the original message never leaves ably so you get all the delivery guarantees
- if the cloud function only sometimes needs to augment messages then it doesn't need to send any additional messages and customers can save on message quota
- architecturally simple, users need pub/sub auth for a single channel to join a chat

Cons:
- if you need to filter messages for inappropriate content then the original objectionable message is still shared with all participants which is not desirable
- the cloud function publishing back on the channel will invoke itself again and the function must ignore messages from itself
- if every message needs to be reformatted and only the augmented message matters then message quota is wasted delivering the original message

##Message filtering

For use cases where you want messages checked, edited or rejected according to a content policy and messages cannot be broadcast to the chat group without being checked.

Users now only have permissions to subscribe to the chat channel and cannot publish to it directly. Instead they publish to separate channel for which they do not have subscribe permissions. This publish channel has a rule configured that invokes the cloud function that performs content checking editing. The function then publishes the checked message to the subscribe only chat channel (it alone has the permissions to do so).

In the case where the function may reject or discard messages then this may need to be fed back to the publisher. The publisher will need a dedicated inbox channel to which the function may publish.

The tricky part is for the function to know which channel name to forward the the filtered message on to or what the inbox channel for the client is. The naive approach would be to include that in the published message but the contents of the inbound message are not trusted so a client could pretend another channel was actually their inbox and we don't have a means with tokens yet to authenticate those claims or pass them on to the function. So I think you have to rely on the inbound channel name and the client ID (as those can be authorised by the token) and have a consistent name mapping for inbound channel -> outbound channel and client ID -> inbox.

Benefits:
- messages that violate the content policy are never broadcast
- only a single broadcast message per publish

Cons:
- more complex, you have to manage multiple different permissions in a token
- going through an external function impacts guaranteed message delivery
- may increase latency on message delivery

**Comment by ben-pickering** [comment]
Is the pattern of using multiple channels to represent one logical object common with Ably? As you point out, it involves some sort of binding between the channels (here it's the naming convention).

Would each user have their own channel to post messages? It would make sense, so that other users don't see the pre-transformed content?

**Comment by cruickshankpg** [comment]
> Would each user have their own channel to post messages? It would make sense, so that other users don't see the pre-transformed content?

there's no need to have per user publish channels, as I said the users do not have subscribe permissions on the publish channel so they can't see pre-transformed content

**Comment by cruickshankpg** [comment]
> Is the pattern of using multiple channels to represent one logical object common with Ably? As you point out, it involves some sort of binding between the channels (here it's the naming convention).

that's probably too broad a question for here but there are plenty of examples of multiple channels being used to create a logical object

**Comment by ben-pickering** [comment]
> there's no need to have per user publish channels, as I said the users do not have subscribe permissions on the publish channel so they can't see pre-transformed content

Ah, so the 'inbox' channel is not the same as the one used for the initial publishing, I see. So for a chat session with N participants, there's N+2 channels: N inboxes, a 'readonly' viewing channel and a publishing channel?

**Comment by cruickshankpg** [comment]
> Ah, so the 'inbox' channel is not the same as the one used for the initial publishing, I see. So for a chat session with N participants, there's N+2 channels: N inboxes, a 'readonly' viewing channel and a publishing channel?

Yes so if there are actually M chat rooms and N total users then there will be 2M + N channels in use
